### PR TITLE
CDAP-4931: Use Path instead of URIs for Artifacts

### PR DESCRIPTION
**Pull request by deepakwadhwani**
Includes fix for Unit Tests. 
Clean Build: http://builds.cask.co/browse/CDAP-DUT5328-1

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 1]
```diff
@@ -1104,23 +1117,86 @@ private ArtifactCell(Id.Artifact artifactId) {
...
+  private static class ArtifactDataDeserializer implements JsonDeserializer<ArtifactData> {
+    @Override
+    public ArtifactData deserialize(JsonElement json,
+                                    Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+      JsonObject jsonObj = json.getAsJsonObject();
+      String locationPath = null;
+      // For Backward Compatibility
+      if (jsonObj.has("locationURI")) {
+        URI locationURI = context.deserialize(jsonObj.get("locationURI"), URI.class);
+        locationPath = locationURI.getPath();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java]: doesn't this have the same problem as before where its not relative?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 2]
```diff
@@ -1104,23 +1117,86 @@ private ArtifactCell(Id.Artifact artifactId) {
...
+    public PluginData deserialize(JsonElement json,
+                                  Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+      JsonObject jsonObj = json.getAsJsonObject();
+      PluginClass pluginClass = context.deserialize(jsonObj.get("pluginClass"), PluginClass.class);
+      ArtifactRange usableBy = context.deserialize(jsonObj.get("usableBy"), ArtifactRange.class);
+      // For Backward Compatibility
+      String artifactLocationPath = null;
+      if (jsonObj.has("artifactLocationURI")) {
+        URI locationURI = context.deserialize(jsonObj.get("artifactLocationURI"), URI.class);
+        artifactLocationPath = locationURI.getPath();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java]: same here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 3]
```diff
@@ -1104,23 +1117,86 @@ private ArtifactCell(Id.Artifact artifactId) {
...
+    @Override
+    public AppData deserialize(JsonElement json,
+                               Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+      JsonObject jsonObj = json.getAsJsonObject();
+      ApplicationClass appClass = context.deserialize(jsonObj.get("appClass"), ApplicationClass.class);
+      // For Backward Compatibility
+      String artifactLocationPath = null;
+      if (jsonObj.has("artifactLocationURI")) {
+        URI locationURI = context.deserialize(jsonObj.get("artifactLocationURI"), URI.class);
+        artifactLocationPath = locationURI.getPath();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java]: and here

**Review by albertshau — CHANGES_REQUESTED** [review]

**Comment by deepakwadhwani** [comment]
Build running: http://builds.cask.co/browse/CDAP-DUT5328-3

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 4]
```diff
@@ -1091,36 +1089,97 @@ private ArtifactCell(Id.Artifact artifactId) {
 
   // Data that will be stored for an artifact. Same as ArtifactDetail, expected without the id since that is redundant.
   private static class ArtifactData {
+    // URI for Backward Compatibility
     private final URI locationURI;
+    private final String locationPath;
     private final ArtifactMeta meta;
 
     ArtifactData(Location location, ArtifactMeta meta) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:1108]: seems like we should only have this constructor

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 5]
```diff
@@ -1091,36 +1089,97 @@ private ArtifactCell(Id.Artifact artifactId) {
...
 
   // Data that will be stored for a plugin.
   private static class PluginData {
     private final PluginClass pluginClass;
     private final ArtifactRange usableBy;
+    // URI for Backward Compatibility
     private final URI artifactLocationURI;
+    private final String artifactLocationPath;
 
     PluginData(PluginClass pluginClass, ArtifactRange usableBy, Location artifactLocation) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:1123]: seems like we should only have this constructor

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 6]
```diff
@@ -1091,36 +1089,97 @@ private ArtifactCell(Id.Artifact artifactId) {
...
   }
 
   // Data that will be stored for an application class.
   private static class AppData {
     private final ApplicationClass appClass;
+    // URI for Backward Compatibility
     private final URI artifactLocationURI;
+    private final String artifactLocationPath;
 
     AppData(ApplicationClass appClass, Location artifactLocation) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:1138]: seems like we should only have this constructor

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java [hunk 7]
```diff
@@ -1091,36 +1089,97 @@ private ArtifactCell(Id.Artifact artifactId) {
...
+    PluginData(PluginClass pluginClass, ArtifactRange usableBy, URI artifactLocationURI) {
       this.pluginClass = pluginClass;
       this.usableBy = usableBy;
-      this.artifactLocationURI = artifactLocation.toURI();
+      this.artifactLocationURI = artifactLocationURI;
+      this.artifactLocationPath = null;
+    }
+
+    private Location getLocation(LocationFactory locationFactory) {
+      Location artifactLocation = this.artifactLocationURI != null
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java]: this logic is repeated 3 times and will be used in other places too. We can move it to the Locations class as something like `Locations.getCompatibleLocation(LocationFactory locationFactory, @Nullable String path, @Nullable URI uri)`

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by deepakwadhwani** [comment]
clean build: http://builds.cask.co/browse/CDAP-DUT5328

**Pull request merged**